lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -2242,9 +2242,47 @@ struct IntelLSUControlsInfo {
   llvm::Optional<unsigned> PrefetchInfo;
 };
 
-IntelFPGADecorations
-tryParseIntelFPGAAnnotationString(StringRef AnnotatedCode) {
-  IntelFPGADecorations Decorates;
+void processOptionalAnnotationInfo(Constant *Const,
+                                   std::string &AnnotationString) {
+  // Handle optional parameter. It can be for example
+  // { %struct.S, i8*, void ()* } { %struct.S undef, i8* null,
+  //                                void ()* @_Z4blahv }
+  // Now we will just handle integer constants (wrapped in a constant
+  // struct, that is being bitcasted to i8*), converting them to string.
+  // TODO: remove this workaround when/if an extension spec that allows or adds
+  // variadic-arguments UserSemantic decoration
+  if (!Const->getNumOperands())
+    return;
+  if (auto *CStruct = dyn_cast_or_null<ConstantStruct>(Const->getOperand(0))) {
+    uint32_t NumOperands = CStruct->getNumOperands();
+    if (!NumOperands)
+      return;
+    if (auto *CInt = dyn_cast<ConstantInt>(CStruct->getOperand(0))) {
+      AnnotationString += ": ";
+      AnnotationString += std::to_string(CInt->getSExtValue());
+    }
+    for (uint32_t I = 1; I != CStruct->getNumOperands(); ++I) {
+      if (auto *CInt = dyn_cast<ConstantInt>(CStruct->getOperand(I))) {
```

Comment:
If some CStruct param is not an integer then it will be just dropped. I.e. '30, "word", 60' will be translated to '30, 60'.
Maybe it would be better to translate optional parameters only when all of them are integers and none of them if there is at least one non integer value.

Reply:
I'd prefer to translate as much as possible within the current approach.

Also, right now it's not a real-life scenario (or at least not something that our backend dev team desire (and since there is no much interest externally I'm guessing that it's not a real-life scenario)). So taking in account, that this code will be replaced with another code within few months I wouldn't bother about it (yet it doesn't bring any bugs/crashes etc).